### ISSUE TITLE
[Feature] how to cal loss when in eval/predict mode

### ISSUE THREAD

**Issue by jeannotes** [open-mmlab/mmengine#881]
### What is the feature?

hi, in basemodel, when mode is not loss, whatever it is val or test, seems all go to "predict" mode, and the result is we cannot get loss when in eval mode, how can we do that?

thanks

**Comment by HAOCHENYE** [comment]
Currently, MMEngine does not support counting average validation loss and calculating metrics at the same time. If you only want to calculate validation loss, maybe you consider these two workarounds

**First method**. Return the validation loss in `predict` mode.
- Return a loss dict in `forward` under `predict` mode
- Set `val_evaluator` as an empty list `[]`
- Set `LogProcessor` as

  ```python
   log_processor = dict(custom_cfg=[dict(data_src='{name of validation loss}', method_name='mean', window_size='epoch')])
   ```

Then, maybe you can see the average validation loss in logs (I've not had a try, maybe it will work).

2. **Second Method**： You can define a `CustomValLoop` to calculate metric and count validation loss at the same time!

**Comment by jeannotes** [comment]
@HAOCHENYE 
1.now we don't have _val_iter, it's hard to log iter when in eval mode, highly recommend add property of val_iter.
I see that when in mode of "after_val_iter" , it just  "logger.ingo(log_str)".
2. highly recommend add val loss and val metric evaluation, this is really useful

**Comment by HAOCHENYE** [comment]
Currently, we use `batch_idx` to represent `_val_iter` for single dataloader training. We'll provide a loop to calculate loss and metric simultaneously ASAP

**Comment by jeannotes** [comment]
@HAOCHENYE thanks for that, and when this is ready, give me a remind
